Add `pass` to removed blocks of code

vthemelis opened the pull request:
Solves #352 more generally.

We saw this in exception handlers but I think that it can manifest itself on any construct that has an expected child block.

We could potentially even totally remove the statements in the test as they are no-op.

agronholm commented:
You need to fix the linting error and add a note to the changelog. What's with the gitignore change?

vthemelis commented:
> You need to fix the linting error

Fixed

>  and add a note to the changelog.

Added

> What's with the gitignore change?

I was getting a lot of `.coverage.<blah> ` files when running `tox` locally. I can remove the change if you prefer but those files are not covered by any of the existing rules and I don't think that they should ever be committed to source control.

agronholm commented on .gitignore:
```diff
@@ -3,6 +3,7 @@
 .idea
 .tox
 .coverage
+**/.coverage.*
```
```suggestion
.coverage*
```

agronholm commented on docs/versionhistory.rst:
```diff
@@ -7,10 +7,14 @@ This library adheres to `Semantic Versioning 2.0 <https://semver.org/#semantic-v
 
 - Dropped Python 3.7 support
 
+- Fixed ``@typechecked`` optimization causing compilation of instrumented code to fail
+  when any block was left empty by the AST transformer (eg `if` or `try` / `except` blocks)
+  (`#352 <https://github.com/agronholm/typeguard/issues/352>`_)
```
```suggestion
- Fixed ``@typechecked`` optimization causing compilation of instrumented code to fail
  when any block was left empty by the AST transformer (eg `if` or `try` / `except` blocks)
  (`#352 <https://github.com/agronholm/typeguard/issues/352>`_)
```

agronholm commented on src/typeguard/_transformer.py:
```diff
@@ -65,6 +66,8 @@
 from dataclasses import dataclass, field
 from typing import Any, ClassVar, cast, overload
 
+from typing_extensions import override
```
Can't have an unconditional import of `typing_extensions` when it's a conditional dependency in `pyproject.toml`. Please remove the `@override` decorators. They're not essential to this change. I'll add them later if I deem them useful.

agronholm replied:
Ignore; I removed it myself.

agronholm commented on src/typeguard/_transformer.py:
```diff
@@ -498,6 +499,21 @@ def __init__(
+
+        # Add `pass` to list fields that were optimised away
+        for field_name in non_empty_list_fields:
+            if not hasattr(node, field_name) or not getattr(node, field_name):
```
```suggestion
            if not getattr(node, field_name, None):
```

agronholm commented:
The test you added seems to already pass in `master`.

agronholm commented:
Thanks!

vthemelis commented:
@agronholm, I opened https://github.com/python/cpython/issues/108521 where they pointed out that maybe we should only be filling in `body` attributes of `AST` nodes. Would you like me to raise another PR to fix this?

agronholm commented:
Sure.
